### PR TITLE
Added basic sessionizer plugin.

### PR DESCRIPTION
**Pull request by nnyx7**
Some changes made to testlib.py to support our testing.

### CODE REVIEW

File: timesketch/lib/analyzers/sessionizer_test.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+
+    @mock.patch(u'timesketch.lib.analyzers.interface.ElasticsearchDataStore',
+                MockDataStore)
+    def test_multiple_sessions(self):
+        """Test multiple events, two of which are in the same session and
+        one in a different session."""
+        with mock.patch.object(SessionizerSketchPlugin,
+                               'event_stream',
+                               return_value=_create_mock_event(
+                                   0, 3, time_diffs=[300000000, 400000000])):
```
Comment (tomchop) [inline timesketch/lib/analyzers/sessionizer_test.py]: I would use a value a bit smaller than `300000000` here so as to make the test more clear (you're already testing the edge case in the test after this)

File: timesketch/lib/analyzers/sessionizer_test.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    ds = MockDataStore("test", 0)
+    user = User('test_user')
+    sketch = Sketch('test_sketch', 'description', user)
+    label = sketch.Label(label='Test label', user=user)
+    sketch.labels.append(label)
+
+    ts = 1410895419859714
+    event_template = ds.get_event('test', 'test')
+
+    for i in range(quantity):
```
Comment (tomchop) [inline timesketch/lib/analyzers/sessionizer_test.py]: I would do some changes to this function to generate a more realistic-looking set of events.

Instead of having `quantity` events separated by `time_diffs`, I would add "batches" of 100 (or any other number) of events, separated by `time_diffs`.

So instead of having (assuming `event_id,timestamp`)

```
event1,0
event2,3000000 # first element of time_diff
event3,4000000 # second element of time_diff
```

I'd have

```
event1,0
event2,1
event3,2
...
event100,3000000 # first element of `time_diff`
event101,3000001
event102,3000002
...
event200,4000000 # second element of time_diff
event201,4000001
```

It shouldn't take much tweaking and it will make the dataset look more realistic, potentially revealing other missed corner-cases.

File: timesketch/lib/analyzers/sessionizer.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+                if curr_time_stamp - last_time_stamp > self.max_time_diff:
+                    session_num += 1
+                event.add_attributes({'session_number': session_num})
+                event.commit()
+
+                last_time_stamp = curr_time_stamp
+
+            self.sketch.add_view('Session view',
+                                 'sessionizer',
+                                 query_string=query)
```
[the remark this answers is not in the document]
Comment (nnyx7) [reply timesketch/lib/analyzers/sessionizer.py]: It looks like it's mandatory to specify query_string or query_dsl, otherwise there's a ValueError.

File: timesketch/lib/analyzers/sessionizer_test.py [hunk 4]
```diff
@@ -231,6 +264,24 @@ def _create_mock_event(event_id, quantity, time_diffs=None):
 
         yield eventObj
 
+        # adding extra events after every requested event for better simulation
+        # of real timeline data i.e. working with a larger dataset
+        for _ in range(100):
+            ts += 1
+            event = event_template
```
Comment (tomchop) [inline timesketch/lib/analyzers/sessionizer_test.py]: This block (up to L 280) is identical to the one in lines 254-263. Maybe extract to a method?

File: timesketch/lib/analyzers/sessionizer_test.py [hunk 5]
```diff
@@ -49,7 +48,9 @@ def test_same_session(self):
             ds = MockDataStore("test", 0)
```
Comment (tomchop) [inline timesketch/lib/analyzers/sessionizer_test.py]: Please use single quotes for strings throughout the codebase.

File: timesketch/lib/analyzers/sessionizer.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+
+class SessionizerSketchPlugin(interface.BaseSketchAnalyzer):
+    """Sessionizing sketch analyser. All events in sketch with id sketch_id
+    are grouped in sessions based on the time difference between them. Two
+    consecutive events are in the same session if the time difference between
+    them is less or equal then max_time_diff_micros"""
+
+    NAME = 'sessionizer'
+    max_time_diff_micros = 300000000
```
Comment (berggren) [inline timesketch/lib/analyzers/sessionizer.py]: I opt for making this configurable. This can wait until the user input to analyzers are ready, but maybe add a TODO so we remember it?

File: timesketch/lib/testlib.py [hunk 7]
```diff
@@ -191,6 +211,34 @@ def create_index(self, *args, **kwargs):
...
+                     event_id=None, flush_interval=None):
+        """Mock adding the event to Elasticsearch, instead add the event
+        to event_store.
+
+        Args:
+            flush_interval: Number of events to queue up before indexing. (This
+            functionality is not supported.)
+            index_name: Name of the index in Elasticsearch
+            event_type: Type of event (e.g. plaso_event)
+            event: Event dictionary
```
Comment (berggren) [inline timesketch/lib/testlib.py]: I would call this event_dict

File: timesketch/lib/testlib.py [hunk 8]
```diff
@@ -191,6 +211,34 @@ def create_index(self, *args, **kwargs):
...
+            event_id: Event Elasticsearch ID
+        """
+
+        for e in self.event_store:
```
Comment (berggren) [inline timesketch/lib/testlib.py]: and instead of "e", call it "event"

File: timesketch/lib/analyzers/sessionizer.py (hunk 6 above)
Comment (katemacleod) [reply timesketch/lib/analyzers/sessionizer.py]: We are planning on changing this in the next pull request, however we will add a to-do.

File: timesketch/lib/testlib.py (hunk 7 above)
Comment (nnyx7) [reply timesketch/lib/testlib.py]: It's also called event in import_event in elastic.py (the function we are mocking). Should we leave this, or change both?

Comment (berggren) [reply timesketch/lib/testlib.py]: Ah, that's right. Then leave it :)

File: timesketch/lib/analyzers/sessionizer.py (hunk 6 above)
Comment (berggren) [reply timesketch/lib/analyzers/sessionizer.py]: Great!
